### PR TITLE
Parser

### PR DESCRIPTION
**Pull request by YaroslavSemeniuk**
Update parser

### CODE REVIEW

File: src/modules/category/category.service.ts [hunk 1]
```diff
@@ -23,22 +27,30 @@ export class CategoryService {
...
+    newCategory.seoId = slugify(data.title);
+    return this.categoryRepository.save(newCategory);
   }
 
   async updateCategory(data: UpdateCategoryDto): Promise<Category> {
     const category = await this.categoryRepository.findOne(data.id);
     if (!category) throw new MessageCodeError('category:notFound');
+    if (data.title) {
+      category.seoId = slugify(data.title);
+    }
```
Comment (olegtsigulov) [inline src/modules/category/category.service.ts]: and if such a seoid already exists? when creating, you check, but not here

File: src/modules/category/category.service.ts [hunk 2]
```diff
@@ -23,22 +27,30 @@ export class CategoryService {
...
 
   async removeCategory(id: string): Promise<boolean> {
+    const categoryExist = await this.categoryRepository.findOne(id);
+    if (!categoryExist) throw new MessageCodeError('category:notFound');
+    const arrayOfIds = [id];
+    const categoryIsUsed = await this.articleRepository.find({ where: { categoriesIds: arrayOfIds } });
+    if (categoryIsUsed.length > 0) throw new MessageCodeError('category:isUsed');
     const deleteResponse = await this.categoryRepository.delete(id);
     return !!deleteResponse.affected;
   }
```
Comment (olegtsigulov) [inline src/modules/category/category.service.ts]: it is not works

File: src/modules/database/entities/article.entity.ts [hunk 3]
```diff
@@ -63,16 +78,15 @@ export class Article {
...
+  @IsArray()
+  categoriesIds: string[]
 
   @ManyToOne(() => User, (user) => user.articles)
   user: User;
 
-  @ManyToOne(() => Category, (category) => category.articles)
-  category: Category;
+  @OneToMany(() => Category, (category) => category.article)
+  categories: Category[];
```
Comment (olegtsigulov) [inline src/modules/database/entities/article.entity.ts]: many to many

File: src/modules/database/entities/category.entity.ts [hunk 4]
```diff
@@ -14,15 +14,32 @@ export class Category {
...
+  })
+  @ApiPropertyOptional({ description: 'category description', example: 'Relationships in the family' })
+  @IsOptional()
+  @IsString()
+  description: string;
 
-  @OneToMany(() => Article, (article) => article.category)
-  articles: Article[]
+  @ManyToOne(() => Article, (article) => article.categories)
+  article: Article
```
Comment (olegtsigulov) [inline src/modules/database/entities/category.entity.ts]: many to many

File: src/modules/database/entities/socialLinks.entity.ts [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+    @ApiPropertyOptional({
+      description: 'link to user account',
+      example: 'https://www.instagram.com/userProfile',
+    })
+    @IsOptional()
+    @IsString()
+    link: string;
+
+    @ManyToOne(() => User, (user) => user.answers, { onDelete: 'CASCADE' })
+    user: User;
```
Comment (olegtsigulov) [inline src/modules/database/entities/socialLinks.entity.ts]: what is it?!?!

File: src/modules/parser/parser.service.ts [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+        private readonly articleRepository: Repository<Article>,
+        @InjectRepository(User)
+        private readonly userRepository: Repository<User>,
+        @InjectRepository(Category)
+        private readonly categoryRepository: Repository<Category>,
+  ) {}
+
+  async getArticlesFromTable():Promise<void> {
+    await this.getArticles();
+  }
```
Comment (olegtsigulov) [inline src/modules/parser/parser.service.ts]: this is a lazy method, I did it with the expectation that you will receive all the information through one method, but if you divided it into different routes, then remove the proxy methods, they are useless

File: src/modules/parser/parser.service.ts [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+        category1, category2, category3, gender, , , authorSeoId,
+      ] = row;
+      const user = await this.userRepository.findOne({ where: { seoId: authorSeoId } });
+      if (!user) continue;
+
+      const categoriesIds = _.compact([category1, category2, category3]);
+      for (const categoryId of categoriesIds) {
+        const existCategory = await this.categoryRepository.findOne({ seoId: categoryId });
+        if (!existCategory) throw new MessageCodeError('category:notFound');
+      }
```
Comment (olegtsigulov) [inline src/modules/parser/parser.service.ts]: dont throw error, push errored column into [ ], and return it to client at final

File: src/modules/parser/parser.service.ts [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+        title,
+        description,
+        descriptionHtml,
+        imgSrc,
+        gender,
+        userId: user.id,
+        categoriesIds,
+      });
+      const errorsList = await validate(object, { whitelist: true });
+      if (errorsList.length) continue;
```
Comment (olegtsigulov) [inline src/modules/parser/parser.service.ts]: push errored column into [ ], and return it to client at final

File: src/modules/parser/parser.service.ts [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+      const existArticle = await this.articleRepository.findOne({ seoId: slugify(title) });
+      if (!existArticle) {
+        await this.articleRepository.save({ ...object, seoId: slugify(title) });
```
Comment (olegtsigulov) [inline src/modules/parser/parser.service.ts]: take out slugify(title) into constant, and use it in this methods

File: src/modules/parser/parser.service.ts [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+      }
+    }
+  }
+
+  async getAuthorsFromTable():Promise<void> {
+    await this.getAuthors();
+  }
```
Comment (olegtsigulov) [inline src/modules/parser/parser.service.ts]: this is a lazy method

File: src/modules/parser/parser.service.ts [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+      const [, name, description, descriptionHtml, imgSrc, socialLinks, gender] = row;
+
+      const object = plainToClass(CreateUserDto, {
+        name, role: 'author', gender, imgSrc, description, descriptionHtml,
+      });
+      const errorsList = await validate(object, { whitelist: true });
+      if (errorsList.length) continue;
+      const existAuthor = await this.userRepository.findOne({ seoId: slugify(name) });
+      if (!existAuthor) {
+        await this.userRepository.save({ ...object, seoId: slugify(name) });
```
Comment (olegtsigulov) [inline src/modules/parser/parser.service.ts]: slugify(name) in constant

File: src/modules/parser/parser.service.ts [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+      }
+    }
+  }
+
+  async getCategoriesFromTable():Promise<void> {
+    await this.getCategories();
+  }
```
Comment (olegtsigulov) [inline src/modules/parser/parser.service.ts]: this is a lazy method

File: src/modules/parser/parser.service.ts [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+
+    for (const row of sheetData) {
+      const [, title, description] = row;
+
+      const object = plainToClass(CreateCategoryDto, { title, description });
+      const errorsList = await validate(object, { whitelist: true });
+      if (errorsList.length) continue;
+      const existCategory = await this.categoryRepository.findOne({ seoId: slugify(title) });
+      if (!existCategory) {
+        await this.categoryRepository.save({ ...object, seoId: slugify(title) });
```
Comment (olegtsigulov) [inline src/modules/parser/parser.service.ts]: +

File: src/modules/user/user.service.ts [hunk 14]
```diff
@@ -1,37 +1,51 @@
...
 
   async createUser(data: CreateUserDto): Promise<User> {
-    const existUser = await this.userRepository.findOne({ email: data.email });
+    const existUser = await this.userRepository.findOne({ name: data.name });
     if (existUser) throw new MessageCodeError('user:exist');
+
     const newUser = this.userRepository.create(data);
-    await this.userRepository.save(data);
-    return newUser;
+    newUser.seoId = slugify(data.name);
```
Comment (olegtsigulov) [inline src/modules/user/user.service.ts]: seoId unique field, await this.userRepository.findOne({ name: data.name }); user or condition here to find by one of fields

File: src/modules/user/user.service.ts [hunk 15]
```diff
@@ -1,37 +1,51 @@
...
+      newUser.socialLinks = await this.socialLinkRepository.save(data.socialLinks);
+    }
+    return this.userRepository.save(newUser);
   }
 
   async updateUser(data: UpdateUserDto): Promise<User> {
     const user = await this.userRepository.findOne(data.id);
     if (!user) throw new MessageCodeError('user:notFound');
+    if (data.name) {
+      user.seoId = slugify(data.name);
```
Comment (olegtsigulov) [inline src/modules/user/user.service.ts]: check seoId for exists?

**Review by olegtsigulov — COMMENTED** [review]
fix it

File: src/modules/database/entities/socialLinks.entity.ts (hunk 5 above)
Comment (YaroslavSemeniuk) [reply src/modules/database/entities/socialLinks.entity.ts:31]: This is the entity that represents the social networks of the user in the form of a record: 
 "socialLinks": [
    {
      "name": "Instagram",
      "link": "https://www.instagram.com/userProfile"
    }

File: src/modules/category/category.service.ts (hunk 1 above)
Comment (YaroslavSemeniuk) [reply src/modules/category/category.service.ts]: done

File: src/modules/database/entities/article.entity.ts (hunk 3 above)
Comment (YaroslavSemeniuk) [reply src/modules/database/entities/article.entity.ts]: done

File: src/modules/database/entities/category.entity.ts (hunk 4 above)
Comment (YaroslavSemeniuk) [reply src/modules/database/entities/category.entity.ts]: done

File: src/modules/database/entities/socialLinks.entity.ts (hunk 5 above)
Comment (YaroslavSemeniuk) [reply src/modules/database/entities/socialLinks.entity.ts]: This is a relationship to the table that stores the user's social links. For exemple:
"socialLinks": [
      {
        "name": "Instagram",
        "link": "https://www.instagram.com/userProfile"
      }
    ]

File: src/modules/parser/parser.service.ts (hunk 6 above)
Comment (YaroslavSemeniuk) [reply src/modules/parser/parser.service.ts]: done

File: src/modules/database/entities/article.entity.ts (hunk 3 above)
Comment (YaroslavSemeniuk) [reply src/modules/database/entities/article.entity.ts]: done

File: src/modules/parser/parser.service.ts (hunk 10 above)
Comment (YaroslavSemeniuk) [reply src/modules/parser/parser.service.ts]: done

File: src/modules/parser/parser.service.ts (hunk 12 above)
Comment (YaroslavSemeniuk) [reply src/modules/parser/parser.service.ts]: done

File: src/modules/database/entities/category.entity.ts (hunk 4 above)
Comment (YaroslavSemeniuk) [reply src/modules/database/entities/category.entity.ts]: done

File: src/modules/parser/parser.service.ts (hunk 9 above)
Comment (YaroslavSemeniuk) [reply src/modules/parser/parser.service.ts]: done

File: src/modules/parser/parser.service.ts (hunk 11 above)
Comment (YaroslavSemeniuk) [reply src/modules/parser/parser.service.ts]: done

File: src/modules/parser/parser.service.ts (hunk 13 above)
Comment (YaroslavSemeniuk) [reply src/modules/parser/parser.service.ts]: done

File: src/modules/database/entities/socialLinks.entity.ts (hunk 5 above)
Comment (YaroslavSemeniuk) [reply src/modules/database/entities/socialLinks.entity.ts]: done

File: src/modules/category/category.service.ts (hunk 1 above)
Comment (YaroslavSemeniuk) [reply src/modules/category/category.service.ts]: done

File: src/modules/user/user.service.ts (hunk 14 above)
Comment (YaroslavSemeniuk) [reply src/modules/user/user.service.ts]: done

File: src/modules/user/user.service.ts (hunk 15 above)
Comment (YaroslavSemeniuk) [reply src/modules/user/user.service.ts]: done

File: src/modules/parser/parser.service.ts (hunk 7 above)
Comment (YaroslavSemeniuk) [reply src/modules/parser/parser.service.ts]: done

File: src/modules/article/article.service.ts [hunk 16]
```diff
@@ -7,75 +7,104 @@ import { MessageCodeError } from '../../shared/errors/message-code-error';
...
 
-    const existCategory = await this.categoryRepository.findOne({ id: data.categoryId });
-    if (!existCategory) throw new MessageCodeError('category:notFound');
+    const categories = [];
+    let categoryId;
+    for (categoryId of data.categoriesIds) {
+      const category = await this.categoryRepository.findOne({ id: categoryId });
+      if (!category) throw new MessageCodeError('category:notFound');
+      categories.push(category);
+    }
```
Comment (olegtsigulov) [inline src/modules/article/article.service.ts]: const categories = await this.categoryRepository.find({ id: data.categoriesIds });
if(categories.length!==data.categoriesIds.length) throw new MessageCodeError('category:notFound');

File: src/modules/article/article.service.ts [hunk 17]
```diff
@@ -7,75 +7,104 @@ import { MessageCodeError } from '../../shared/errors/message-code-error';
...
+      const articleExist = !!await this.articleRepository.findOne({ seoId });
+      if (articleExist) throw new MessageCodeError('article:exist');
+      article.seoId = seoId;
+    }
+    if (data.categoriesIds) {
+      let categoryId;
+      for (categoryId of data.categoriesIds) {
+        const existCategory = await this.categoryRepository.findOne({ id: categoryId });
+        if (!existCategory) throw new MessageCodeError('category:notFound');
+      }
```
Comment (olegtsigulov) [inline src/modules/article/article.service.ts]: const categories = await this.categoryRepository.find({ id: data.categoriesIds });
if(categories.length!==data.categoriesIds.length) throw new MessageCodeError('category:notFound');

File: src/modules/article/article.service.ts [hunk 18]
```diff
@@ -7,75 +7,104 @@ import { MessageCodeError } from '../../shared/errors/message-code-error';
...
     if (!article) throw new MessageCodeError('article:notFound');
+    if (data.title) {
+      const seoId = ToTranslit(data.title);
+      const articleExist = !!await this.articleRepository.findOne({ seoId });
+      if (articleExist) throw new MessageCodeError('article:exist');
+      article.seoId = seoId;
+    }
+    if (data.categoriesIds) {
+      let categoryId;
+      for (categoryId of data.categoriesIds) {
```
Comment (olegtsigulov) [inline src/modules/article/article.service.ts]: what is it!??!? let categoryId

File: src/modules/article/article.service.ts [hunk 19]
```diff
@@ -7,75 +7,104 @@ import { MessageCodeError } from '../../shared/errors/message-code-error';
...
 
-  async getArticlesByGenderAndCategory(gender: string, categoryId: string): Promise<Article[]> {
-    return this.articleRepository.find({
-      where: { gender, categoryId },
-    });
+  async getArticlesByGenderAndCategoryId(gender: string, categoryId: string): Promise<Article[]> {
+    return this.articleRepository.createQueryBuilder('article')
+      .leftJoinAndSelect('article.categories', 'category')
+      .where('category.id = :id', { id: categoryId })
+      .andWhere('article.gender IN (:...genders)', { genders: [gender, 'female both male'] })
```
Comment (olegtsigulov) [inline src/modules/article/article.service.ts]: 'female both male' why didn't you put genders into constants?

File: src/modules/article/dto/received/create-article.dto.ts [hunk 20]
```diff
@@ -43,9 +44,8 @@ export class CreateArticleDto {
     @IsUUID('4')
     userId: string;
 
-    @ApiProperty({ description: 'category by article', example: uuidv4() })
+    @ApiProperty({ description: 'categories by article', example: [uuidv4()] })
     @IsNotEmpty()
-    @IsString()
-    @IsUUID('4')
-    categoryId: string;
+    @IsArray()
```
Comment (olegtsigulov) [inline src/modules/article/dto/received/create-article.dto.ts:48]: @IsUUID('4', { each: true })
    @MinLength(1)
    @IsArray()

File: src/modules/article/dto/received/get-by-category.dto.ts [hunk 21]
```diff
@@ -1,11 +1,10 @@
 import { v4 as uuidv4 } from 'uuid';
 import { ApiProperty } from '@nestjs/swagger';
-import { IsNotEmpty, IsString, IsUUID } from 'class-validator';
+import { IsNotEmpty, IsString } from 'class-validator';
 
 export class GetByCategoryDto {
   @ApiProperty({ description: 'category id', example: uuidv4() })
   @IsNotEmpty()
   @IsString()
-  @IsUUID('4')
```
Comment (olegtsigulov) [inline src/modules/article/dto/received/get-by-category.dto.ts]: why do you remove it?

File: src/modules/article/dto/received/get-by-gender-and-category.dto.ts [hunk 22]
```diff
@@ -15,6 +13,5 @@ export class GetByGenderAndCategoryDto {
     @ApiProperty({ description: 'category id', example: uuidv4() })
     @IsNotEmpty()
     @IsString()
-    @IsUUID('4')
```
Comment (olegtsigulov) [inline src/modules/article/dto/received/get-by-gender-and-category.dto.ts]: +

File: src/modules/parser/parser.controller.ts [hunk 23]
```diff
@@ -0,0 +1,49 @@
...
+    summary: 'Get categories from table',
+    description: 'Save categories from Google SpreadSheets to DB',
+  })
+  @ApiResponse({
+    status: HttpStatus.OK,
+    description: 'Data obtained',
+  })
+  getCategoriesFromTable(): Promise<any> {
+    return this.parserService.getCategories();
+  }
```
Comment (olegtsigulov) [inline src/modules/parser/parser.controller.ts]: Promise<any>

File: src/modules/parser/parser.service.ts [hunk 24]
```diff
@@ -0,0 +1,131 @@
...
+        continue;
+      }
+      const seoId = ToTranslit(title);
+      const category = await this.categoryRepository.findOne({ seoId });
+      if (!category) {
+        await this.categoryRepository.save({ ...object, seoId });
+      } else unsavedCategories.push(category.seoId);
+    }
+    unsavedCategories.push(_.compact(validationError));
+    return unsavedCategories;
```
Comment (olegtsigulov) [inline src/modules/parser/parser.service.ts]: unsavedAuthors.push({ seoId: clientSeoId, reason: 'User already exists' });

**Review by olegtsigulov — COMMENTED** [review]
fix it

File: src/modules/article/dto/received/get-by-category.dto.ts (hunk 21 above)
Comment (YaroslavSemeniuk) [reply src/modules/article/dto/received/get-by-category.dto.ts]: done

File: src/modules/article/dto/received/create-article.dto.ts (hunk 20 above)
Comment (YaroslavSemeniuk) [reply src/modules/article/dto/received/create-article.dto.ts:48]: done

File: src/modules/article/article.service.ts (hunk 19 above)
Comment (YaroslavSemeniuk) [reply src/modules/article/article.service.ts]: done

File: src/modules/article/article.service.ts (hunk 18 above)
Comment (YaroslavSemeniuk) [reply src/modules/article/article.service.ts]: done

File: src/modules/article/article.service.ts (hunk 17 above)
Comment (YaroslavSemeniuk) [reply src/modules/article/article.service.ts]: done

File: src/modules/article/article.service.ts (hunk 16 above)
Comment (YaroslavSemeniuk) [reply src/modules/article/article.service.ts]: done

File: src/modules/article/dto/received/get-by-gender-and-category.dto.ts (hunk 22 above)
Comment (YaroslavSemeniuk) [reply src/modules/article/dto/received/get-by-gender-and-category.dto.ts]: done

**Review by olegtsigulov — APPROVED** [review]

**Pull request merged**